[Part] feature to make persistent section cut views

donovaly opened the pull request:
* one aim is to have a feature to make real cuts, not hollow ones like with clipping
* the second one is to have a persistent view, so an object in the tree that can be further exported, meshed etc.

For a detailed description with examples see https://forum.freecadweb.org/viewtopic.php?f=17&t=52441

chennes commented:
@donovaly is this still a WIP? It sounded from the forums discussion like you were going to make some changes to it.

donovaly commented:
This is still WIP. As discussed the feature is useful for many users but it needs to be polished:
- better name to make clear the cut is persistent
- the sliders must be removed since they will freeze FC for larger bodies
- some bugfixes, for example, when closing the dialog, the cuts should be default disappear, but this is not yet working for all cases.

chennes commented:
Thanks -- if it's not ready for review yet, can you mark it as a draft? Thanks!

donovaly commented:
> Thanks -- if it's not ready for review yet, can you mark it as a draft? Thanks!

I already marked it as "[WIP]" since Draft would be the WB.
I would also set a label but I don't know how this is done.

chennes commented:
Sorry I wasn't clear: I didn't mean to edit the title, but to mark its status as draft. It prevents it from being committed, and is easy to filter. See https://docs.github.com/en/github/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request

donovaly commented:
> See https://docs.github.com/en/github/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request

I cannot do this since I don't have the permissions. I need the same permissions needed to set e.g. a label or milestone.

chennes commented:
Interesting, I thought the owner could always do it, but I guess only when creating the PR? Anyway, I took care of it now. I assume you can see the "Ready for review" button at the bottom?

donovaly commented:
> I took care of it now. I assume you can see the "Ready for review" button at the bottom?

Yes. I pressed it now and then I see the option to make it as draft. So it seems I can mark the PR as draft as PR creator. So I will take care of this in future.

berndhahnebach commented:
Following a link to the branch on the CI-repository:

https://gitlab.com/berndhahnebach/FreeCAD/-/commits/PR_5060

The CI-status is available on the latest commit of the branch.
If there is no status available the PR should be rebased on latest master.
Check pipeline branches to see if your PR has been run by the CI.

https://gitlab.com/berndhahnebach/FreeCAD/-/pipelines?scope=branches

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_4118"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_4118/pipeline.svg"/></a> for feature branch [PR_4118](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_4118). Pipeline [460535341](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/460535341) was triggered at [662b2f0](https://github.com/FreeCAD/FreeCAD/pull/4118/commits/662b2f04e3f19539113e76816eea30ea2a2f3d7c). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_4118"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_4118/pipeline.svg"/></a> for feature branch [PR_4118](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_4118). Pipeline [461515823](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/461515823) was triggered at [413b119](https://github.com/FreeCAD/FreeCAD/pull/4118/commits/413b119c8e7ee9fe00d68705377756dfc025e847). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

donovaly commented:
This PR is in my opinion now ready.

chennes commented on src/Mod/Part/Gui/SectionCutting.cpp:
```diff
@@ -0,0 +1,1479 @@
+    }
+    // the case of an empty box and having cuts will be handles later by startCutting(true)
+
+    connect(ui->groupBoxX, SIGNAL(toggled(bool)), this, SLOT(onGroupBoxXtoggled(bool)));
```
I suggest migrating this to the non-macro versions so that you get type-checking. So, e.g. something like
```suggestion
    connect(ui->groupBoxX, &QGroupBox::toggled, this, &SectionCut::onGroupBoxXtoggled);
```

donovaly replied:
Thanks.

Done now. I see that I am the first one in the whole code using 
`QGroupBox::toggled`

I did what you suggested but I am not sure I did it right for QDoubleSpinBox. I found this:
https://doc.qt.io/qt-5/qdoublespinbox.html#valueChanged
Can you please have a Look if I implemented this correctly?

chennes replied:
It looks right to me: the nice thing about the new style is that if you are doing it wrong you get a compiler error instead of having to wait for it to fail at runtime. So mostly if it compiles it's correct. I have not actually tested the interface yet, but I think you've got the code alright now.

donovaly replied:
Thanks. Yes, it compiles without warnings, the CI run fine and I checked that all intended functionality is there.

Since the UI was in the making for more than a year now, it is now time to go live. -> merging

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_4118"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_4118/pipeline.svg"/></a> for feature branch [PR_4118](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_4118). Pipeline [462378305](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/462378305) was triggered at [1c99b47](https://github.com/FreeCAD/FreeCAD/pull/4118/commits/1c99b47899793d11b5493a310f1c7519a9e2b7d7). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

wwmayer commented on src/Mod/Part/Gui/SectionCutting.cpp:
```diff
@@ -0,0 +1,1475 @@
+SbBox3f SectionCut::getViewBoundingBox()
+{
+    auto docGui = Gui::Application::Instance->activeDocument();
+    Gui::View3DInventor* view = static_cast<Gui::View3DInventor*>(docGui->getActiveView());
```
@donovaly The static_cast is unsafe because the active view can be something else, e.g. a spreadsheet view. Use a dynamic_cast and check the result.

wwmayer commented on src/Mod/Part/Gui/SectionCutting.cpp:
```diff
@@ -0,0 +1,1475 @@
+void SectionCut::onRefreshCutPBclicked()
+{
+    // get document
+    doc = Gui::Application::Instance->activeDocument()->getDocument();
```
@donovaly This check is unsafe. Because if there is not active document then `activeDocument()` already returns null and thus will cause a segmentation fault.

donovaly replied:
done in #6360

wwmayer commented on src/Mod/Part/Gui/SectionCutting.cpp:
```diff
@@ -0,0 +1,1475 @@
+    Gui::View3DInventorViewer* viewer = view->getViewer();
+
+    // get all objects in the document
+    doc = Gui::Application::Instance->activeDocument()->getDocument();
```
@donovaly This check is unsafe. Because if there is not active document then `activeDocument()` already returns null and thus will cause a segmentation fault.

donovaly replied:
done in #6360

wwmayer commented on src/Mod/Part/Gui/SectionCutting.cpp:
```diff
@@ -0,0 +1,1475 @@
+        return;
+    }
+    // empty the ObjectsListVisible
+    ObjectsListVisible.clear();
```
ObjectsListVisible cannot be non-empty here.

donovaly replied:
I don't understand. It must be cleared here, otherwise it won't work.

wwmayer replied:
We are inside the constructor of SectionCut and at this point no member function will be called to modify the class member ObjectsListVisible. Since it's empty by default how can it become non-empty here?

donovaly replied:
Ah, my mistake. I thought I have to initialize it explicitly as being empty to assure that there is no garbage in it when it is not initialized.
Since it already is initialized as empty, I can omit this.
(seems I programmed recently too much in FreePascal ;-) )

donovaly replied:
done in #6360

wwmayer commented on src/Mod/Part/Gui/SectionCutting.cpp:
```diff
@@ -0,0 +1,1475 @@
+        }
+        ui->cutZ->setValue(pcBox->Height.getValue() - fabs(pcBox->Placement.getValue().getPosition().z));
+    }
+    if (doc->getObject(BoxYName)) {
```
Same as above

wwmayer commented on src/Mod/Part/Gui/SectionCutting.cpp:
```diff
@@ -0,0 +1,1475 @@
+    if (doc->getObject(BoxXName)) {
+        hasBoxX = true;
+        ui->groupBoxX->setChecked(true);
+        Part::Box* pcBox = static_cast<Part::Box*>(doc->getObject(BoxXName));
```
Same as above

donovaly replied:
done in #6360

wwmayer commented on src/Mod/Part/Gui/SectionCutting.cpp:
```diff
@@ -0,0 +1,1475 @@
+    if (doc->getObject(BoxYName)) {
+        hasBoxY = true;
+        ui->groupBoxY->setChecked(true);
+        Part::Box* pcBox = static_cast<Part::Box*>(doc->getObject(BoxYName));
```
Same as above

donovaly replied:
done in #6360

wwmayer commented on src/Mod/Part/Gui/SectionCutting.cpp:
```diff
@@ -0,0 +1,1475 @@
+    if (doc->getObject(CutZName)) {
+        anObject = doc->getObject(CutZName);
+        // the deleted object might have been visible before, thus check and delete it from the list
+        for (it = ObjectsListVisible.begin(); it != ObjectsListVisible.end(); ++it) {
```
You better use std::find:
```
auto it = std::find(ObjectsListVisible.begin(), ObjectsListVisible.end(), anObject);
if (it != ObjectsListVisible.end())
    ObjectsListVisible.erase(it);
```

donovaly replied:
done in #6360

wwmayer commented on src/Mod/Part/Gui/SectionCutting.cpp:
```diff
@@ -0,0 +1,1475 @@
+    }
+    if (doc->getObject(BoxZName)) {
+        anObject = doc->getObject(BoxZName);
+        for (it = ObjectsListVisible.begin(); it != ObjectsListVisible.end(); ++it) {
```
Same as above

wwmayer commented on src/Mod/Part/Gui/SectionCutting.cpp:
```diff
@@ -0,0 +1,1475 @@
+    }
+    if (doc->getObject(CutYName)) {
+        anObject = doc->getObject(CutYName);
+        for (it = ObjectsListVisible.begin(); it != ObjectsListVisible.end(); ++it) {
```
Same as above

wwmayer commented on src/Mod/Part/Gui/SectionCutting.cpp:
```diff
@@ -0,0 +1,1475 @@
+    }
+    if (doc->getObject(BoxYName)) {
+        anObject = doc->getObject(BoxYName);
+        for (it = ObjectsListVisible.begin(); it != ObjectsListVisible.end(); ++it) {
```
Same as above

wwmayer commented on src/Mod/Part/Gui/SectionCutting.cpp:
```diff
@@ -0,0 +1,1475 @@
+    }
+    if (doc->getObject(CutXName)) {
+        anObject = doc->getObject(CutXName);
+        for (it = ObjectsListVisible.begin(); it != ObjectsListVisible.end(); ++it) {
```
Same as above

wwmayer commented on src/Mod/Part/Gui/SectionCutting.cpp:
```diff
@@ -0,0 +1,1475 @@
+        std::vector<App::DocumentObject*> compoundObjects;
+        pcCompoundDel->Links.getLinks(compoundObjects);
+        // first delete the compound
+        for (it = ObjectsListVisible.begin(); it != ObjectsListVisible.end(); ++it) {
```
Same as above

wwmayer commented on src/Mod/Part/Gui/SectionCutting.cpp:
```diff
@@ -0,0 +1,1475 @@
+    }
+    if (doc->getObject(CompoundName)) {
+        auto compoundObject = doc->getObject(CompoundName);
+        Part::Compound* pcCompoundDel = static_cast<Part::Compound*>(compoundObject);
```
Bad static_cast

donovaly replied:
done in #6360

wwmayer commented on src/Mod/Part/Gui/SectionCutting.cpp:
```diff
@@ -0,0 +1,1475 @@
+        startCutting();
+        return;
+    }
+    Part::Box* pcBox = static_cast<Part::Box*>(CutBox);
```
Bad static_cast

wwmayer commented on src/Mod/Part/Gui/SectionCutting.cpp:
```diff
@@ -0,0 +1,1475 @@
+        startCutting();
+        return;
+    }
+    Part::Cut* pcCut = static_cast<Part::Cut*>(CutObject);
```
Bad static_cast

wwmayer commented on src/Mod/Part/Gui/SectionCutting.cpp:
```diff
@@ -0,0 +1,1475 @@
+    // to uncheck cutting planes and to add/remove objects while this dialog is open
+    // We must remove in this order because the tree hierary of the features is Z->Y->X and Cut->Box
+    App::DocumentObject* anObject;
+    if (doc->getObject(CutZName)) {
```
The code block below appears 6x inside the function (only the name is different). Write a lambda function to avoid code duplication.

donovaly replied:
done in #6360

donovaly commented on src/Mod/Part/Gui/SectionCutting.cpp:
```diff
@@ -0,0 +1,1475 @@
+    if (doc->getObject(BoxZName)) {
+        hasBoxZ = true;
+        ui->groupBoxZ->setChecked(true);
+        Part::Box* pcBox = static_cast<Part::Box*>(doc->getObject(BoxZName));
```
But in the next line I check if it is a Part::Box. What is my mistake?

wwmayer replied:
You assume that an existing object "SectionCutBoxZ" must be a Part::Box because you create an object in the dialog with this (default) name. But you ignore the fact that the user may have created an object with this name that is not a Part::Box but e.g. a Mesh::Cube.

Doing a static_cast on a pointer is _not_ a type check. The static_cast assumes that you know what you are doing and simply returns a pointer to a Part::Box. In case the object is a Part::Box it works but silently fails if it's not. For the latter case you still get a pointer but you cannot trust it because it's already a dangling pointer.

There are two way to solve this:
* replace the static_cast with a dynamic_cast. This cast internally checks the class' typeid and if it doesn't match it will return a nullptr.
* before doing a static_cast you must use the function `isDerivedFrom(Part::Box::getClassTypeId())`. Internally this uses FreeCAD's implementation of type ids.

donovaly replied:
> Doing a static_cast on a pointer is not a type check.

Aha! Lesson learned. I thought this is the case.
@chennes since you also had a look at this code, it might be a valuable info for you too.

donovaly replied:
done in #6360

donovaly commented:
Many thanks for the review!

Not that you get the impression that I force-merged it. The point was that this PR was here fore ore than a year and 9 months as normal PR. Meanwhile I got frequent user requests to finish it and this way forced myself to eventually polish it. You once already reviewed it and I thought I changed the code already accordingly. @chennes also had a look and I changed what he objected.

I will try to fix all things you mentioned on which I did not yet answered. For those I answered to, I need more info or help.

I am currently abroad and cannot code. I will be back next week.

I have one request and some info:
1. I got positive feedback from users and even videos for the feature, despite it is clear that it needs more work. But I am now happy since without feedback it is hard to know what to implement in what way, where do users misunderstand something etc. I see that for some use cases, the feature is not self-explanatory, but that is acceptable as I got as feedback. I will setup a Wiki page to explain the feature properly. Then I found a bug meanwhile by myself on a real-life use case. And I also got a bug report.
2. This PR you merged today, breaks the feature: #6213 When one cannot open the Section cut dialog from the View menu while e.g. being in an assembly WB, the concept of the feature fails and also the usability would not be there. I commented in that PR and ask you to have a look.
3. The feature needs in my opinion 2 more things:
A. An option to all ow cutting objects overlapping each other. In this case I need to put th objects to be cut into a Boolean Fragments container and not into a compound. I already had a look but could not find a C++ way on how to create a BoleanFragments object which I can will with other objects. Do you have a pointer for me where to look?
B. the cut faces should get a special color. This is what most feedback persons told me. I have an idea and will try to make a PR that does this.

By the way, is there a reason why the Color per face feature is only available from within the Part WB? I was recently asked about this since people want to color the faces the Section cut feature created. I think it should be available from within PartDesign and all assembly WBs.

donovaly commented on src/Mod/Part/Gui/SectionCutting.cpp:
```diff
@@ -0,0 +1,1475 @@
+    // make all objects visible that have been visible when the dialog was called
+    // because we made them invisible when we created cuts
+    for (it = ObjectsListVisible.begin(); it != ObjectsListVisible.end(); ++it) {
+        if ((*it)->isValid()) // a formerly visible object might have been deleted
```
To understand: How can there be dangling pointers? I mean I just created this list few lines above. And on creation time I checked that the objects exist. Even if the users wants, he cannot have deleted the objects within the few milliseconds between the list creation and checking.

Nevertheless I can change to App::DocumentObjectT .

wwmayer replied:
You fill the list ObjectsListVisible in the constructor of SectionCut. But between the constructor call and this function call a lot of things can happen because the dialog is non-modal and thus allows the user to modify the document in the meantime.

Assume you added three objects to ObjectsListVisible and the user has deleted one of them before continuing the interaction with the dialog. Now the dialog class has nothing implemented to be notified about document changes and thus won't update ObjectsListVisible accordingly.

donovaly replied:
done in #6360

wwmayer commented:
> Not that you get the impression that I force-merged it.

I don't mind. Things can still be fixed afterwards.

> The point was that this PR was here fore ore than a year and 9 months as normal PR.

The PR was in a pending state from my last review on and I did't have a look any more afterwards.

> You once already reviewed it and I thought I changed the code already accordingly

Parts of it but not everything, e.g. the std::find stuff. In general using std::find has some advantages over using a for-loop:
* for a reviewer it's immediately clear what the intention is while with a for-loop its body must be studied first (Not a big improvement in this case)
* the code will be smaller (Not a big improvement in this case)
* it's faster and safer

If you use a for-loop like this:
`for (it = ObjectsListVisible.begin(); it != ObjectsListVisible.end(); ++it) {`
then for every iteration step it must evaluate ObjectsListVisible.end(). Additionally it can happen that you unintentionally remove an element from the vector while iterating it and this invalidates the iterator it and thus will cause a crash for the next iteration step.
Now have a look at https://en.cppreference.com/w/cpp/algorithm/find where you can find its implementation:
```
template<class InputIt, class T>
constexpr InputIt find(InputIt first, InputIt last, const T& value)
{
    for (; first != last; ++first) {
        if (*first == value) {
            return first;
        }
    }
    return last;
}
```
In the for-loop it uses first != last and thus doesn't have to create a tmp. iterator object like it does with ObjectsListVisible.end().

Btw, when changing ObjectsListVisible to use DocumentObjectT you cannot use std::find any more but you have to use std::find_if instead. Your code will then look like:
```
        App::DocumentObject* anObject = doc->getObject(CutZName);
        std::vector<App::DocumentObjectT> ObjectsListVisible;
        auto it = std::find_if(ObjectsListVisible.begin(), ObjectsListVisible.end(), [anObject](const App::DocumentObjectT& obj) {
            return (obj.getObject() == anObject);
        });
        if (it != ObjectsListVisible.end()) {
            ObjectsListVisible.erase(it);
        }
```

> I got positive feedback from users

My feedback is not about the provided functionality, it's only about unsafe code that partially leads to hard to find crashes.

> This PR you merged today, breaks the feature:

I answered you there...

> A. An option to allow cutting objects overlapping each other.

It's neither tested nor compiled code but the idea is to do something like this:
```
PyObject* module = PyImport_ImportModule("BOPTools.SplitFeatures");
if (!module) {
    // handle exception
    return;
}

std::string name = ...
Py::Module split(module, true);
Py::Object result(split.callMemberFunction("makeBooleanFragments", Py::TupleN(Py::String(name))));

if (PyObject_TypeCheck(result.ptr(), &Part::PartFeaturePy::Type)) {
    Part::Feature* part = static_cast<Part::FeaturePy*>(result.ptr())->getFeaturePtr();
    App::PropertyLinkList* prop = dynamic_cast<App::PropertyLinkList*>(part->getPropertyByName("Objects"));
    if (prop) {
...
    }
...
}
```

> By the way, is there a reason why the Color per face feature is only available from within the Part WB?

It's also accessible from within PD in the context-menu.

donovaly commented:
> > By the way, is there a reason why the Color per face feature is only available from within the Part WB?
> 
> It's also accessible from within PD in the context-menu.

How and where? I cannot select a face and then change its color via the context menu.

I added now the face coloring to the View menu like it is in Part:
/commit/9b1245ab5

donovaly commented:
I changed the code as suggested in #6360. Can you please have a look if I did it right.
(It compiles and works so far, I was also able to trigger some cases were the new error messages occur, so where I would have before invalid memory access.)
